src/Features/CSharp/Portable/ConvertPrimaryToRegularConstructor/ConvertPrimaryToRegularConstructorCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,505 @@
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics.Analyzers.NamingStyles;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Indentation;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertPrimaryToRegularConstructor;
+
+using static SyntaxFactory;
+
+[ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = PredefinedCodeRefactoringProviderNames.ConvertPrimaryToRegularConstructor), Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed partial class ConvertPrimaryToRegularConstructorCodeRefactoringProvider()
+    : CodeRefactoringProvider
+{
+    public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+    {
+        var (document, span, cancellationToken) = context;
+        var typeDeclaration = await context.TryGetRelevantNodeAsync<TypeDeclarationSyntax>().ConfigureAwait(false);
+        if (typeDeclaration?.ParameterList is null)
+            return;
+
+        // Converting a record to a non-primary-constructor form is a lot more work (for example, having to synthesize a
+        // Deconstruct method, and figure out how to specify properties, etc.).  We can consider adding support for that
+        // scenario later if desired.
+        if (typeDeclaration is RecordDeclarationSyntax)
+            return;
+
+        var triggerSpan = TextSpan.FromBounds(typeDeclaration.SpanStart, typeDeclaration.ParameterList.FullSpan.End);
+        if (!triggerSpan.Contains(span))
+            return;
+
+        context.RegisterRefactoring(CodeAction.Create(
+                CSharpFeaturesResources.Convert_to_regular_constructor,
+                cancellationToken => ConvertAsync(document, typeDeclaration, typeDeclaration.ParameterList, context.Options, cancellationToken),
+                nameof(CSharpFeaturesResources.Convert_to_regular_constructor)),
+            triggerSpan);
+    }
+
+    private static async Task<Solution> ConvertAsync(
+        Document document,
+        TypeDeclarationSyntax typeDeclaration,
+        ParameterListSyntax parameterList,
+        CodeActionOptionsProvider optionsProvider,
+        CancellationToken cancellationToken)
+    {
+        var compilation = await document.Project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
tried to doc this method to help make every rewrite step clear.  let me know if more clarity is needed.